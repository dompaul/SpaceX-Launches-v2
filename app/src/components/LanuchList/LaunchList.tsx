import React from "react";
import { LaunchItem } from "../LaunchItem";
import { LaunchOption } from "../../models/LaunchModel";
import LABEL from "../../constants/Labels";

interface LaunchListProps {
  items: LaunchOption[];
  filter: string;
  loaded: boolean;
  error: boolean;
  testId?: string;
  sort: boolean;
}

export const LaunchList: React.FC<LaunchListProps> = ({
  items,
  filter,
  testId,
  loaded,
  error,
  sort,
}) => {
  let filteredItems = [...items];

  if (error) {
    return <span data-testid="error-state-test">{LABEL.ERROR}</span>;
  }

  if (!loaded) {
    return <span data-testid="loading-state-test">{LABEL.LOADING}</span>;
  }

  if (filter !== "") {
    filteredItems = [...items].filter((item: LaunchOption) => {
      const year = Number(new Date(item.launch_date_utc).getFullYear());
      return year === Number(filter);
    });
  }

  const itemsCopy = filteredItems.sort((a, b) => {
    const x = Number(new Date(a.launch_date_utc).getTime());
    const y = Number(new Date(b.launch_date_utc).getTime());
    return sort ? y - x : x - y;
  });

  return (
    <ul className="launch-list" data-testid={testId}>
      {itemsCopy.map((item, index) => {
        return <LaunchItem key={index} item={item} index={index} />;
      })}
    </ul>
  );
};
