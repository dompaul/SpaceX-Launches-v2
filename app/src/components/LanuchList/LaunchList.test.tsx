import { getByTestId, render, RenderResult } from "@testing-library/react";
import LABEL from "../../constants/Labels";
import { LaunchList } from "./LaunchList";

const testIds = {
  error: "error-state-test",
  load: "loading-state-test",
  launchList: "launch-list-test",
};

const mockLaunchList = [
  {
    flight_number: "1",
    mission_name: "FalconSat",
    launch_date_utc: "2006-03-24T22:30:00.000Z",
    rocket: {
      rocket_name: "Falcon 1",
    },
  },
  {
    flight_number: "2",
    mission_name: "RazakSat",
    launch_date_utc: "2009-07-13T03:35:00.000Z",
    rocket: {
      rocket_name: "Falcon 1",
    },
  },
];

describe("LaunchList.tsx", () => {
  let wrapper: RenderResult;

  describe("error", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchList
          error={true}
          items={mockLaunchList}
          filter="2009"
          loaded={true}
          sort={true}
        ></LaunchList>
      );
    });

    it("api fails", () => {
      const message = getByTestId(wrapper.container, testIds.error);
      expect(message).toHaveTextContent(LABEL.ERROR);
    });
  });

  describe("load", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchList
          error={false}
          items={mockLaunchList}
          filter="2009"
          loaded={false}
          sort={true}
        ></LaunchList>
      );
    });

    it("api still loading", () => {
      const load = getByTestId(wrapper.container, testIds.load);
      expect(load).toHaveTextContent(LABEL.LOADING);
    });
  });

  describe("filter", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchList
          error={false}
          items={mockLaunchList}
          filter="2009"
          loaded={true}
          sort={true}
          testId={testIds.launchList}
        ></LaunchList>
      );
    });

    it("launch list is filtered", () => {
      const launchList = getByTestId(wrapper.container, testIds.launchList);
      expect(launchList.childElementCount).toEqual(1);
    });
  });

  describe("sort ascending order", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchList
          error={false}
          items={mockLaunchList}
          filter=""
          loaded={true}
          sort={false}
          testId={testIds.launchList}
        ></LaunchList>
      );
    });

    it("launch list is sorted ascending order", () => {
      expect(
        getByTestId(wrapper.container, testIds.launchList)
      ).toMatchSnapshot();
    });
  });

  describe("sort descending order", () => {
    beforeEach(() => {
      wrapper = render(
        <LaunchList
          error={false}
          items={mockLaunchList}
          filter=""
          loaded={true}
          sort={true}
          testId={testIds.launchList}
        ></LaunchList>
      );
    });

    it("launch list is sorted descending order", () => {
      expect(
        getByTestId(wrapper.container, testIds.launchList)
      ).toMatchSnapshot();
    });
  });
});
