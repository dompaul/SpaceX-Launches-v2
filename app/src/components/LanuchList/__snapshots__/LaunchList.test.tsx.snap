// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchList.tsx sort ascending order launch list is sorted ascending order 1`] = `
<ul
  class="launch-list"
  data-testid="launch-list-test"
>
  <li
    class="launch-item"
  >
    <div
      class="launch-item__mission-content"
    >
      <span
        class="launch-item__number"
      >
        #1
      </span>
      <span
        class="launch-item__label"
      >
        FalconSat
      </span>
    </div>
    <div
      class="launch-item__meta-info"
    >
      <span
        class="launch-item__date"
      >
        <time
          datetime="1143239400000"
        >
          24th March 2006
        </time>
      </span>
      <span
        class="launch-item__rocket"
      >
        Falcon 1
      </span>
    </div>
  </li>
  <li
    class="launch-item"
  >
    <div
      class="launch-item__mission-content"
    >
      <span
        class="launch-item__number"
      >
        #2
      </span>
      <span
        class="launch-item__label"
      >
        RazakSat
      </span>
    </div>
    <div
      class="launch-item__meta-info"
    >
      <span
        class="launch-item__date"
      >
        <time
          datetime="1247456100000"
        >
          13th July 2009
        </time>
      </span>
      <span
        class="launch-item__rocket"
      >
        Falcon 1
      </span>
    </div>
  </li>
</ul>
`;

exports[`LaunchList.tsx sort descending order launch list is sorted descending order 1`] = `
<ul
  class="launch-list"
  data-testid="launch-list-test"
>
  <li
    class="launch-item"
  >
    <div
      class="launch-item__mission-content"
    >
      <span
        class="launch-item__number"
      >
        #2
      </span>
      <span
        class="launch-item__label"
      >
        RazakSat
      </span>
    </div>
    <div
      class="launch-item__meta-info"
    >
      <span
        class="launch-item__date"
      >
        <time
          datetime="1247456100000"
        >
          13th July 2009
        </time>
      </span>
      <span
        class="launch-item__rocket"
      >
        Falcon 1
      </span>
    </div>
  </li>
  <li
    class="launch-item"
  >
    <div
      class="launch-item__mission-content"
    >
      <span
        class="launch-item__number"
      >
        #1
      </span>
      <span
        class="launch-item__label"
      >
        FalconSat
      </span>
    </div>
    <div
      class="launch-item__meta-info"
    >
      <span
        class="launch-item__date"
      >
        <time
          datetime="1143239400000"
        >
          24th March 2006
        </time>
      </span>
      <span
        class="launch-item__rocket"
      >
        Falcon 1
      </span>
    </div>
  </li>
</ul>
`;
